[pack] extension bundle probing and download

soninaren opened the pull request:
Resolves #3711
Resolves #3961
Resolves #3714
Resolves #4061

soninaren commented:
Created a new PR with same content from #4043.

mattchenderson commented:
@soninaren What's the status here? Is there anything we're waiting on before reviewers can get going?

mattchenderson commented:
@soninaren what are the chances we get this into this next release?

soninaren commented:
> @soninaren What's the status here? Is there anything we're waiting on before reviewers can get going?

PR is ready for review. I have addressed all the comments requested in the last iteration.

> @soninaren what are the chances we get this into this next release?
Need quick feedback loop from @fabiocav , @brettsam , @pragnagopa. The PR has gone through a number of iterations. So chances are pretty good for getting this checked in.

pragnagopa commented on src/WebJobs.Script.WebHost/Controllers/ExtensionsController.cs:
```diff
@@ -58,6 +61,10 @@ public async Task<IActionResult> Get()
         {
+            if (_extensionBundleManager.IsExtensionBundleConfigured())
+            {
+                return BadRequest();
```
Update to return BadRequest with an error message: https://docs.microsoft.com/en-us/dotnet/api/system.web.http.apicontroller.badrequest?view=aspnetcore-2.2#System_Web_Http_ApiController_BadRequest_System_String_

pragnagopa commented on src/WebJobs.Script.WebHost/Controllers/ExtensionsController.cs:
```diff
@@ -98,7 +105,7 @@ public async Task<IActionResult> GetJobs()
-            if (package == null)
+            if (package == null || _extensionBundleManager.IsExtensionBundleConfigured())
             {
                 return BadRequest();
```
Same here. Split package == null and _extensionBundleManager.IsExtensionBundleConfigured().  Throw BadRequest with a message if extensionbundle is configured

pragnagopa commented on src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs:
```diff
@@ -48,16 +54,32 @@ private static string[] GetBuiltinExtensionAssemblies()
 
         public Type[] GetStartupTypes()
```
Is this used anywhere?

soninaren replied:
Yes, This is called by the webjobs. When the host starts up it uses the instance of ScriptStartupTypeLocator provided by us and calls the GetStartupTypes() to locate the extensions.

pragnagopa commented on src/WebJobs.Script/ExtensionBundle/ExtensionBundleManager.cs:
```diff
@@ -0,0 +1,189 @@
+        {
+            bundlePath = null;
+            string bundleMetatdataFile = null;
+            var paths = new List<string>(_options.ProbingPaths)
```
move this to constructor

soninaren replied:
It is only used for this method and is a temporary construct. Do we want to add as a instance variable and alive for the entire lifetime of the object?

soninaren commented on src/WebJobs.Script.WebHost/WebScriptHostBuilderExtension.cs:
```diff
@@ -26,6 +26,11 @@ public static class WebScriptHostBuilderExtension
+            var hostingEnvironment = rootServiceProvider.GetService<AspNetCore.Hosting.IHostingEnvironment>();
+            if (hostingEnvironment != null)
+            {
+                builder.Properties.Add(nameof(AspNetCore.Hosting.IHostingEnvironment), hostingEnvironment);
```
It is used within the ExtensionBundleOptionsSetup to identify if we are running in development environment? Based on that i am setting the correct downloadPath and ProbingPath in ExtensionBundleOptionsSetup.

soninaren commented on src/WebJobs.Script/ExtensionBundle/ExtensionBundleManager.cs:
```diff
@@ -0,0 +1,189 @@
+
+        public async Task<string> GetExtensionBundle(HttpClient httpClient = null)
+        {
+            if (httpClient == null)
```
Actually this is done on purpose based on the feedback from @fabiocav. That is how it was done before. Using the httpClient from DI container would create an HttpClient and keep it scope for the entire life time of the host. Eventually we want to remove the instance of the HttpClient in the DI container.

soninaren commented on src/WebJobs.Script/ExtensionBundle/ExtensionBundleManager.cs:
```diff
@@ -0,0 +1,189 @@
+            return !string.IsNullOrEmpty(_options.Id) && !string.IsNullOrEmpty(_options.Version?.OriginalString);
+        }
+
+        public async Task<string> GetExtensionBundle(HttpClient httpClient = null)
```
It actually downloads the extensionBundle if extension bundle is not present and returns the bundle path for the downloaded extensionBundle. GetExtensionBundlePath kind of masks all the operations happening behind the scenes and simple suggests that the method is fetching path for an existing bundle. However i do feel we need to capture that the method does indeed return a string which is the bundle path.

pragnagopa replied:
GetExtensionBundlePath makes sense. Please add summary comments for this method, that explains  bundle is downloaded if not present

pragnagopa commented on src/WebJobs.Script/ExtensionBundle/ExtensionBundleManager.cs:
```diff
@@ -0,0 +1,189 @@
+            else
+            {
+                return await GetBundle(httpClient);
+            }
```
Remove Method GetBundle. Move logic to new up HttpClient if null to GetExtensionBundle.

soninaren replied:
can't new up the HttpClient in GetBundle method. The httpClient is inside using statement in one instance vs the other one is plainly using it. Removing GetBundle method. I would have to copy the entire logic inside if block as well as the else block statement.

fabiocav replied:
To have a simpler API,  you should have two distinct overloads of GetExtensionBundle, the one that does not take an `HttpClient` would dispose of the client it creates.

pragnagopa commented on src/WebJobs.Script/ScriptConstants.cs:
```diff
@@ -122,6 +122,13 @@ public static class ScriptConstants
+        public const string CdnBaseUri = "https://functionscdn.azureedge.net/public";
+        public const string ExtensionBundleMetadatFile = "bundle.json";
+        public const string ExtensionBundleVersionIndexFile = "index.json";
+        public const string LogCategoryExtensionBundles = "Host.ExtensionBundles";
```
Instead of adding log category here, use CreateLogger<T>

pragnagopa commented on src/WebJobs.Script/ExtensionBundle/ExtensionBundleManager.cs:
```diff
@@ -0,0 +1,189 @@
+        public ExtensionBundleManager(ExtensionBundleOptions options, IEnvironment environment, ILoggerFactory loggerFactory)
+        {
+            _environment = environment ?? throw new ArgumentNullException(nameof(environment));
+            _logger = loggerFactory.CreateLogger(ScriptConstants.LogCategoryExtensionBundles) ?? throw new ArgumentNullException(nameof(loggerFactory));
```
change to _logger = loggerFactory.CreateLogger<ExtensionBundleManager>();

soninaren replied:
Adding `loggerFactory.CreateLogger<ExtensionBundleManager>` based on the discussion.

pragnagopa commented on src/WebJobs.Script.WebHost/Controllers/ExtensionsController.cs:
```diff
@@ -58,6 +62,11 @@ public async Task<IActionResult> Get()
         public async Task<IActionResult> Delete(string id)
         {
+            if (_extensionBundleManager.IsExtensionBundleConfigured())
+            {
```
As discussed in person, please add tests

soninaren replied:
There aren't any tests cases for any of the controller yet. We might need to setup something in e2e tests. Have added test cases for changes to extension manager in the extension manager tests.

pragnagopa commented on test/WebJobs.Script.Tests/ScriptStartupTypeDiscovererTests.cs:
```diff
@@ -20,7 +23,7 @@ namespace Microsoft.Azure.WebJobs.Script.Tests
     {
         [Fact]
-        public void GetExtensionsStartupTypes_FiltersBuiltinExtensions()
+        public async System.Threading.Tasks.Task GetExtensionsStartupTypes_FiltersBuiltinExtensionsAsync()
```
Nit: Task

pragnagopa commented on src/WebJobs.Script/Config/ExtensionBundleOptions.cs:
```diff
@@ -2,14 +2,22 @@
+
+        public string DownloadPath { get; set; }
+
+        public bool EnsureLatest { get; set; }
```
Nit: Please add Summary for these properties as only Id and version can user controlled.

fabiocav commented on src/WebJobs.Script/ExtensionBundle/IExtensionBundleManager.cs:
```diff
@@ -0,0 +1,15 @@
+{
+    public interface IExtensionBundleManager
+    {
+        Task<string> GetExtensionBundle(HttpClient httpClient = null);
```
Based on the comment above, just have two overloads of this method, with an without the client argument

fabiocav commented on test/WebJobs.Script.Tests/Configuration/ExtensionBundleOptionsSetupTest.cs:
```diff
@@ -171,5 +257,22 @@ private IConfiguration BuildHostJsonConfiguration(IEnvironment environment = nul
             return configurationBuilder.Build();
         }
+
+        private TestEnvironment GetAppServiceEnvironment()
```
Can we have a better name for this method? The current name doesn't clearly describe what it does.

fabiocav commented on test/WebJobs.Script.Tests/ExtensionBundleManagerTests.cs:
```diff
@@ -0,0 +1,301 @@
+
+            protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+            {
+                await Task.Delay(1);
```
Can you clarify? May want to just use `Task.Yield` instead

soninaren replied:
Just used it to make Task.Dely. Task.Yield works as well.

soninaren commented on src/WebJobs.Script/Config/ExtensionBundleOptionsSetup.cs:
```diff
@@ -0,0 +1,93 @@
+
+                if (_environment.IsAppServiceEnvironment() || _hostingEnvironment.IsDevelopment())
+                {
+                    options.DownloadPath = Path.Combine(_environment.GetEnvironmentVariable(EnvironmentSettingNames.AzureWebsiteHomePath),
```
Non AppService environment would have to override this configuration. For example core tools running in windows would set this to be a different location than core tools running in linux.

soninaren commented on src/WebJobs.Script.WebHost/Controllers/ExtensionsController.cs:
```diff
@@ -184,6 +198,8 @@ private string GetJobBasePath()
                 basePath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".azurefunctions", "extensions");
             }
+
+            FileUtility.EnsureDirectoryExists(basePath);
```
Have added check to verify if persistent file system is available to the calling method.

pragnagopa commented on src/WebJobs.Script/ExtensionBundle/ExtensionBundleManager.cs:
```diff
@@ -39,21 +39,28 @@ public bool IsExtensionBundleConfigured()
+        /// <returns>Path of the extension bundle</returns>
+        public async Task<string> GetExtensionBundlePath(HttpClient httpClient)
+        {
+            return await GetBundle(httpClient);
```
Just return GetBundle(httpClient); Calling function will await

pragnagopa commented on src/WebJobs.Script/ExtensionBundle/ExtensionBundleManager.cs:
```diff
@@ -39,21 +39,28 @@ public bool IsExtensionBundleConfigured()
-            else
+            using (var httpClient = new HttpClient())
             {
                 return await GetBundle(httpClient);
```
Same here Just return GetBundle(httpClient); Calling function will await

pragnagopa commented on src/WebJobs.Script/ScriptHostBuilderExtensions.cs:
```diff
@@ -256,10 +255,10 @@ internal static void ConfigureApplicationInsights(HostBuilderContext context, IL
         {
             var options = new ExtensionBundleOptions();
-            var optionsSetup = new ExtensionBundleOptionsSetup(context.Configuration, SystemEnvironment.Instance, hostingEnvironment);
+            var optionsSetup = new ExtensionBundleOptionsSetup(context.Configuration, SystemEnvironment.Instance, context.HostingEnvironment);
```
👍

fabiocav replied:
It's a bit confusing that this is an `IOptionsSetup<ExtensionBundleOptions>` but not really used as one (not registered, just explicitly used here). I can see that confusing others if they're working on this code. We might want to just drop that implementation and rename this to reflect that this is a configuration helper.

fabiocav commented:
@soninaren just ping when this is ready for review

fabiocav commented on src/WebJobs.Script/Environment/EnvironmentSettingNames.cs:
```diff
@@ -53,6 +53,6 @@ public static class EnvironmentSettingNames
         public const string CoreToolsEnvironment = "FUNCTIONS_CORETOOLS_ENVIRONMENT";
 
-        public const string AlternateCdnUri = "FUNCTIONS_ALTERNATE_CDN_URI";
+        public const string FunctionsFallbackCdnUri = "FUNCTIONS_FALLBACK_CDN_URI";
```
Rename this as discussed.
